src/cpp/src/lm_encoding.cpp

```diff
@@ -213,26 +204,29 @@ std::pair<EncodedResults, int32_t> get_lm_encoded_results(
         raw_perf_counters.m_new_token_times.emplace_back(infer_end);
         raw_perf_counters.m_batch_sizes.emplace_back(batch_size);
 
-        stream_generated_tokens();
-
         sampler_output = sampler.sample(active_sequence_groups, m_llm.get_tensor("logits"));
-
-        active_sequence_groups.erase(std::remove_if(active_sequence_groups.begin(),
-                                                    active_sequence_groups.end(),
-                                                    get_active_sequence_groups),
-                                    active_sequence_groups.end());
+        stream_generated_tokens();
     }
 
-    // to stream last token
-    stream_generated_tokens();
-    if (streamer_ptr) {
+    if (streamer_ptr) { // push streamer's cache
         streamer_ptr->end();
     }
-    
+
+    // Collect results
+
     size_t next_selected_beam = 0;
     for (size_t i = 0; i < sequence_groups.size(); i++) {
         auto request = sequence_groups[i];
-        auto generation_outputs = generations[i]->read_all();
+        std::vector<GenerationOutput> generation_outputs;
+        auto sampling_params = request->get_sampling_parameters();
+
+        for (auto sequence : request->get_sequences()) {
+            GenerationOutput output;
+            output.generated_ids = sequence->get_generated_ids();
+            output.score = sampling_params.is_beam_search() ? sequence->get_beam_search_score(sampling_params) : sequence->get_cumulative_log_probs();
+
+            generation_outputs.push_back(output);
+        }
```

Comment:
can't get_finished_sequences() be taken here? generation_outputs will be already sorted

Reply:
fixed

adopted `get_finished_sequences` to treat sequence groups dropped by handle, as finished